replica/distributed_loader.cc

```diff
@@ -465,22 +462,34 @@ distributed_loader::process_upload_dir(distributed<replica::database>& db, distr
         };
         process_sstable_dir(directory, flags).get();
 
-        auto generation = highest_generation_seen(directory).get0();
-        auto shard_generation_base = generation.value_or(replica::table::make_new_generation()).value() / smp::count + 1;
-
-        // We still want to do our best to keep the generation numbers shard-friendly.
-        // Each destination shard will manage its own generation counter.
-        std::vector<std::atomic<sstables::generation_type::int_t>> shard_gen(smp::count);
-        for (shard_id s = 0; s < smp::count; ++s) {
-            shard_gen[s].store(shard_generation_base * smp::count + s, std::memory_order_relaxed);
-        }
-
-        // we need generation calculated by instance of cf at requested shard
-        auto new_generation_for_shard = [&] (shard_id shard) {
-            auto gen = shard_gen[shard].fetch_add(smp::count, std::memory_order_relaxed);
-            return sstables::generation_type(gen);
+        class generation_generator : public seastar::peering_sharded_service<generation_generator> {
+            sstables::sstable_generation_generator _gen;
+        public:
+            generation_generator(int64_t last_gen)
+                : _gen{last_gen} {}
+            future<sstables::generation_type> next(shard_id shard) {
+                if (shard == this_shard_id()) {
+                    co_return _gen(shard);
+                }
+                co_return co_await container().invoke_on(shard, [shard](auto& gg) {
+                    return gg.next(shard);
+                });
+            }
+        };
```

Comment:
This could be much simpler.
Just have `sharded<sstables::sstable_generation_generator> sharded_generator` and in `make_sstable()` below, you call `auto generation = sharded_generator.invoke_on(shard, [] { ... ]).get();`.

Reply:
thanks! incorporated the proposed change.